File: realtime-macos/Sources/PAPAluLive/CharacterRuntime.swift
enum CharacterDisplayState: Equatable {
    case idle
    case talking
}

struct CharacterRuntime {
    private(set) var definition: CharacterDefinition
    private(set) var state: CharacterDisplayState = .idle
    private var talkingIndex = 0

    var currentAssetName: String {
        switch state {
        case .idle:
            return definition.idleAssetName
        case .talking:
            return definition.talkingAssetNames[talkingIndex]
        }
    }

    init(definition: CharacterDefinition) {
        precondition(!definition.talkingAssetNames.isEmpty)
        self.definition = definition
    }

    mutating func setState(_ state: CharacterDisplayState) {
        self.state = state
        if state == .talking {
            talkingIndex = 0
        }
    }

    mutating func advanceTalkingFrame() {
        guard state == .talking else { return }
        talkingIndex = (talkingIndex + 1) % definition.talkingAssetNames.count
    }

    mutating func setCharacter(
        _ definition: CharacterDefinition,
        currentState: CharacterDisplayState
    ) {
        self.definition = definition
        talkingIndex = 0
        setState(currentState)
    }
}

File: realtime-windows/PAPAluLive.Core/CharacterRuntime.cs
namespace PAPAluLive.Core;

public sealed class CharacterRuntime
{
    private int talkingIndex;

    public CharacterDefinition Definition { get; private set; }
    public CharacterDisplayState State { get; private set; } =
        CharacterDisplayState.Idle;

    public string CurrentAssetName => State switch
    {
        CharacterDisplayState.Idle => Definition.IdleAssetName,
        CharacterDisplayState.Talking => Definition.TalkingAssetNames[talkingIndex],
        _ => throw new InvalidOperationException("Unknown display state."),
    };

    public CharacterRuntime(CharacterDefinition definition)
    {
        if (definition.TalkingAssetNames.Count == 0)
        {
            throw new ArgumentException(
                "A character needs at least one talking asset.",
                nameof(definition));
        }

        Definition = definition;
    }

    public void SetState(CharacterDisplayState state)
    {
        State = state;
        if (state == CharacterDisplayState.Talking)
        {
            talkingIndex = 0;
        }
    }

    public void AdvanceTalkingFrame()
    {
        if (State != CharacterDisplayState.Talking)
        {
            return;
        }

        talkingIndex = (talkingIndex + 1) % Definition.TalkingAssetNames.Count;
    }

    public void SetCharacter(
        CharacterDefinition definition,
        CharacterDisplayState currentState)
    {
        if (definition.TalkingAssetNames.Count == 0)
        {
            throw new ArgumentException(
                "A character needs at least one talking asset.",
                nameof(definition));
        }

        Definition = definition;
        talkingIndex = 0;
        SetState(currentState);
    }
}
